### PR TITLE
Add VCPU IMSIC migration support.

### PR DESCRIPTION
**Pull request by rajnesh-kanwal**
Adds support to rebind a VCPU to a different PCPU without MRIF support. This only supports GIF to GIF migration.

The main idea is to update the IMSIC mappings in guest stage tables and IOMMU to point to new guest interrupt file. This allow new file to capture interrupt while the migration is in process. Given is the overall workflow:

> TvmCpuRebindImsicBegin: Sets up new interrupt file mapping and begins removal of previous one.
> 
> TvmInitiateFence: The host must complete a TLB invalidation sequence to make sure the old mappings are invalidated and new mappings are used.
> 
> TvmCpuRebindImsicClone: Must run on the old PCPU. This clones the old guest interrupt file internally to transfer it to new guest interrupt file. This also checks for TLB invalidation and frees the old guest interrupt file for future use.
> 
> TvmCpuRebindImsicEnd: Copies over the saved guest interrupt file state into the new interrupt file and marks the VCPU state again to Bound.

I must admit I don't have the pages and memory management abstraction fully figured out in my head nice and clean. I feel like there can be better ways of doing the page remapping work. Feel free to point out any problems. Also, new to rust but I have tried my best to not do any silly mistakes. Just 15 days ago, rust looked like gibberish and now I am doing the same gibberish, although not of that quality.

Marking this PR as draft as this is just to gather initial thoughts and there is some cleanup to do in the existing code as well.
- Remove the iommu msi page table related code as without MRIF, there is no point of msi page table.
- Update a couple of TH-API comments. 
- Add (per-CPU!) bitmap/state to keep track of used and unused guest interrupt files to avoid host trying to assign same file to multiple VCPUs.

### CODE REVIEW

File: riscv-page-tables/src/page_table.rs
Comment (dgreid) [inline riscv-page-tables/src/page_table.rs:897]: Because Paddr could point anywhere, I'm not convinced this can be a safe function.

File: riscv-page-tables/src/pte.rs
Comment (dgreid) [inline riscv-page-tables/src/pte.rs:89]: This also seems unsafe, @abrestic-rivos is the 'set' function above still safe?

**Review by dgreid — COMMENTED** [review]
Looks like you're picking up rust very quickly. A few nits. I didn't get to the API, I'll defer to abrestic for that review. Thanks!

File: src/vm.rs
Comment (abrestic-rivos) [inline src/vm.rs:2023]: Need comment about why `unwrap()` is ok here.

File: src/vm_cpu.rs
Comment (abrestic-rivos) [inline src/vm_cpu.rs:1229]: `Rebinding` here and below.

File: src/vm_interrupts.rs
Comment (abrestic-rivos) [inline src/vm_interrupts.rs:27]: Pre-clone and post-clone `Rebinding` should be separate states so that we can enforce that the host does the clone step.

File: drivers/src/imsic/core.rs
Comment (rajnesh-kanwal) [reply drivers/src/imsic/core.rs:661]: Makes sense. I will revert the change.

Comment (rajnesh-kanwal) [reply drivers/src/imsic/core.rs:661]: Fixed.

File: drivers/src/iommu/msi_page_table.rs
Comment (rajnesh-kanwal) [reply drivers/src/iommu/msi_page_table.rs:181]: Fixed.

File: riscv-page-tables/src/page_table.rs
Comment (rajnesh-kanwal) [reply riscv-page-tables/src/page_table.rs:238]: This is actually inspired from `map_leaf`. I kind of found it fine given at this level of stack, if alignment is not valid then that's most probably a programming bug.
I can change both if you want.

File: src/vm_interrupts.rs
Comment (rajnesh-kanwal) [reply src/vm_interrupts.rs:28]: Fixed.

File: drivers/src/iommu/msi_page_table.rs
Comment (rajnesh-kanwal) [reply drivers/src/iommu/msi_page_table.rs]: Fixed.

File: drivers/src/imsic/core.rs
Comment (rajnesh-kanwal) [reply drivers/src/imsic/core.rs]: Fixed.

File: riscv-page-tables/src/page_table.rs
Comment (rajnesh-kanwal) [reply riscv-page-tables/src/page_table.rs]: Code no longer present.

File: riscv-page-tables/src/page_table.rs
Comment (rajnesh-kanwal) [reply riscv-page-tables/src/page_table.rs]: Code no longer present.

File: riscv-page-tables/src/page_table.rs
Comment (rajnesh-kanwal) [reply riscv-page-tables/src/page_table.rs]: Code no longer present.

File: riscv-page-tables/src/pte.rs
Comment (rajnesh-kanwal) [reply riscv-page-tables/src/pte.rs]: Fixed.

File: src/main.rs
Comment (rajnesh-kanwal) [reply src/main.rs]: Fixed by @glg-rv in https://github.com/rivosinc/salus/pull/177.

File: riscv-page-tables/src/page_table.rs
Comment (rajnesh-kanwal) [reply riscv-page-tables/src/page_table.rs]: Fixed.

File: src/vm_pages.rs
Comment (rajnesh-kanwal) [reply src/vm_pages.rs]: Fixed.

File: src/vm_interrupts.rs
Comment (rajnesh-kanwal) [reply src/vm_interrupts.rs]: Fixed.

File: src/vm_interrupts.rs
Comment (rajnesh-kanwal) [reply src/vm_interrupts.rs]: Fixed.

File: drivers/src/imsic/core.rs
Comment (rajnesh-kanwal) [reply drivers/src/imsic/core.rs]: Thanks, sounds good. Fixed.

File: src/vm.rs
Comment (rajnesh-kanwal) [reply src/vm.rs]: Code no longer present.

File: src/vm_cpu.rs
Comment (rajnesh-kanwal) [reply src/vm_cpu.rs]: Fixed.

File: src/vm_interrupts.rs
Comment (abrestic-rivos) [inline src/vm_interrupts.rs:29]: If we're in the `Cloned` state, we've already read out the interrupt state from the previous CPU and the interrupt file is now free to be reused. Thus I don't think it's necessary to hold onto the old CPU / interrupt file in this state.

File: src/vm_interrupts.rs
Comment (abrestic-rivos) [inline src/vm_interrupts.rs:198]: See comment above: `Rebinding` should be the only state where we care about the previous location.

File: riscv-page-tables/src/page_table.rs
Comment (dgreid) [reply riscv-page-tables/src/page_table.rs:238]: If the code calling it has any way to manage the error, then I think it should return an error. Handling a salus panic on a system is going to be rather ugly.

File: riscv-page-tables/src/page_table.rs
Comment (dgreid) [inline riscv-page-tables/src/page_table.rs:570]: is there an extra 'not' in this comment? Calling update_page_addr(paddr) should cause vaddr to point to paddr, or am I lost?

File: test-workloads/src/bin/tellus.rs
Comment (abrestic-rivos) [inline test-workloads/src/bin/tellus.rs:461]: typo in commit message: "rebinding"

**Review by abrestic-rivos — APPROVED** [review]
LGTM thanks, I'll let Dylan give it another look before merging.

**Review by dgreid — APPROVED** [review]
LGTM as well. Thanks!

File: test-workloads/src/bin/tellus.rs
Comment (rajnesh-kanwal) [reply test-workloads/src/bin/tellus.rs:461]: Fixed.

**Comment by rajnesh-kanwal** [comment]
Thanks for the review. @dgreid @abrestic-rivos

**Pull request merged**
